libraries/AP_Airspeed/AP_Airspeed_RSC.cpp

```diff
@@ -0,0 +1,443 @@
+    _set_data_rate(N_DR_175_SPS);
+
+    // convert temperature first
+    _adc_convert(TEMPERATURE);
+
+    // release semaphore
+    _adc_dev->get_semaphore()->give();
+
+    // timer run at 90hz
+    _adc_dev->register_periodic_callback(1000000UL/90U,
+                                     FUNCTOR_BIND_MEMBER(&AP_Airspeed_RSC::_timer, void));
+    return true;
+}
+
+void AP_Airspeed_RSC::_timer()
+{
+    uint8_t buf[3];
+
+    if (!_data_ready()) {
+        return;
+    }
+
+    // read adc value
+    _adc_read(buf);
+
+    WITH_SEMAPHORE(sem);
+
+    if (_type == TEMPERATURE) {
+        // tempreature is 14 bits in 2s complement representation
+        // each byte is shifted to its position in a 16-bit unsigned integer and from 8 more bits to be left-aligned in a 16-bit integer
+        _t_raw = ((uint16_t)buf[0] << 8) | (uint16_t)buf[1];
+        // right-shift signed integer back to get measurement value
+        _t_raw >>= 2;
+
+        // apply scaler to get temperature
+        _temperature = _t_raw * TEMPERATURE_RATIO;
+    } else {
+        // pressure is 24 bits in 2s complement representation
+        // each byte is shifted to its position in a 32-bit unsigned integer and from 8 more bits to be left-aligned in a 32-bit integer
+        _p_raw = ((uint32_t)buf[0] << 24) | ((uint32_t)buf[1] << 16) | ((uint32_t)buf[2] << 8);
+        // right-shift signed integer back to get measurement value
+        _p_raw >>= 8;
+        // apply scaler to get pressure
+        if (_unit == INH2O) {
+            _pressure_sum += _calculate_pressure() * INCH_OF_H2O_TO_PASCAL;
+        }
+        _press_count++;
+        _last_sample_time_ms = AP_HAL::millis();
+    }
+
+    _state = (_state + 1) % 5;
+
+    if (_state == 0) {
+        _adc_convert(TEMPERATURE);
+    } else {
+        _adc_convert(PRESSURE);
+    }
+}
+
+bool AP_Airspeed_RSC::_data_ready()
+{
+    if (_drdy_pin) {
+        return _drdy_pin->read() != 1;
+    }
+
+    return false;
+}
+
+// return the current differential_pressure in Pascal
+bool AP_Airspeed_RSC::get_differential_pressure(float &pressure)
+{
+    WITH_SEMAPHORE(sem);
+
+    if ((AP_HAL::millis() - _last_sample_time_ms) > 100) {
+        return false;
+    }
+
+    if (_press_count > 0) {
+        _pressure = _pressure_sum / _press_count;
+        _press_count = 0;
+        _pressure_sum = 0;
+    }
+
+    pressure = _pressure;
+    return true;
+}
+
+// return the current temperature in degrees C, if available
+bool AP_Airspeed_RSC::get_temperature(float &temperature)
+{
+    WITH_SEMAPHORE(sem);
+
+    if ((AP_HAL::millis() - _last_sample_time_ms) > 100) {
+        return false;
+    }
+
+    temperature = _temperature;
+    return true;
+}
+
+void AP_Airspeed_RSC::_get_pressure_range()
+{
+    uint8_t buf[RSC_PRESSURE_RANGE_LEN];
+    _eeprom_read(RSC_PRESSURE_RANGE_LSB, RSC_PRESSURE_RANGE_LEN, buf);
+    // convert byte array to float (buf[0] is LSB)
+    memcpy(&_pressure_range, &buf, sizeof(_pressure_range));
```

Comment:
Why do you need the `memcpy` here? It seems like this could just read directly into the target buffer, at which point the whole helper could probably be removed (on many of these)

Reply:
let me see it